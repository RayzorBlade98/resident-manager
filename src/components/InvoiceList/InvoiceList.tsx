/* eslint-disable react/jsx-props-no-spreading */
import { StyleAttribute } from 'glamor';
import React from 'react';
import GenericList from '../GenericComponents/GenericList/GenericList';
// eslint-disable-next-line max-len
import GenericListElement from '../GenericComponents/GenericList/GenericListElement';
import { Invoice } from '_/types/invoice';

interface InvoiceListProps {
  /**
   * List of included invoices
   */
  invoices: Invoice[];

  /**
   * Currently selected invoice
   */
  selectedInvoice?: Invoice;

  /**
   * Callback when selecting an invoice
   */
  onSelectInvoice: (invoice: Invoice) => void;

  /**
   * Additional stlye of the list container
   */
  style?: StyleAttribute;
}

/**
 * Component that displays a list of provided invoices
 */
function InvoiceList(props: InvoiceListProps): JSX.Element {
  return (
    <GenericList style={props.style} className="invoiceList">
      {props.invoices.map((invoice: Invoice) => (
        <GenericListElement
          onClick={() => {
            props.onSelectInvoice(invoice);
          }}
          selected={invoice === props.selectedInvoice}
          key={`$invoice-${invoice.start.month}-${invoice.start.year}`}
        >
          {`${invoice.start.month} ${invoice.start.year}`}
        </GenericListElement>
      ))}
    </GenericList>
  );
}

export default InvoiceList;
