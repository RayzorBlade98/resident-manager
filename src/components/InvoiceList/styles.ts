import { css, Rule } from 'glamor';

const newInvoiceElementStyle: Rule = { height: '5%' };

const styles = {
  newInvoiceElement: css(newInvoiceElementStyle),
};

export default styles;
