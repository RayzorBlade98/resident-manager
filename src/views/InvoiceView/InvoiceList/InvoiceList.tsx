import {
  Box,
  Divider,
  List,
  ListItemButton,
  ListItemText,
} from '@mui/material';
import React, { useState } from 'react';
import { useRecoilState, useRecoilValue } from 'recoil';
import { invoiceViewSelectedInvoiceSelector } from '../states/invoice_view_state';
import CreateInvoiceModal from './CreateInvoiceModal/CreateInvoiceModal';
import Invoice from '_/models/invoice/invoice';
import invoiceState from '_/states/invoice/invoice.state';

const styles = {
  box: {
    height: '100%',
  },
  list: {
    height: '100%',
    overflowY: 'auto',
    padding: 0,
    borderRight: 1,
    borderColor: 'divider',
  },
  listItemButton: {
    height: '11.03%',
  },
};

/**
 * Component that displays a list of all invoices
 */
function InvoiceList(): JSX.Element {
  const [showModal, setShowModal] = useState(false);

  const invoices = useRecoilValue(invoiceState);
  const [selectedInvoice, setSelectedInvoice] = useRecoilState(
    invoiceViewSelectedInvoiceSelector,
  );

  return (
    <>
      <CreateInvoiceModal showModal={showModal} onCloseModal={() => setShowModal(false)} />
      <Box sx={styles.box}>
        <List sx={styles.list}>
          <ListItemButton
            onClick={() => setShowModal(true)}
            sx={styles.listItemButton}
          >
            <ListItemText primary="Neue Abrechnung" />
          </ListItemButton>
          <Divider />
          {invoices.map((invoice: Invoice, i) => (
            <>
              <ListItemButton
                selected={invoice.id === selectedInvoice?.id}
                onClick={() => setSelectedInvoice(invoice)}
                key={invoice.id}
                sx={styles.listItemButton}
              >
                <ListItemText
                  primary={`${invoice.start.toString()} - ${invoice.end.toString()}`}
                />
              </ListItemButton>
              {i !== invoices.length - 1 && <Divider />}
            </>
          ))}
        </List>
      </Box>
    </>
  );
}

export default InvoiceList;
