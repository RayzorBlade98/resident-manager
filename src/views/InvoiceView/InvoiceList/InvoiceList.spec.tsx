import {
  RenderResult, act, fireEvent, render,
} from '@testing-library/react';
import { range } from 'lodash';
import React from 'react';
import { getRecoil, setRecoil } from 'recoil-nexus';
import invoiceViewState, {
  invoiceViewSelectedInvoiceSelector,
} from '../states/invoice_view_state';
import InvoiceList from './InvoiceList';
import invoiceState from '_/states/invoice/invoice.state';
import ReactTestWrapper from '_/test/ReactTestWrapper';
import InvoiceBuilder from '_/test/builders/invoice.builder';

describe('InvoiceList', () => {
  const invoices = range(0, 5).map((i) => new InvoiceBuilder().withId(`invoice${i}`).build());
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(
      <ReactTestWrapper>
        <InvoiceList />
      </ReactTestWrapper>,
    );

    act(() => {
      setRecoil(invoiceState, invoices);
      setRecoil(invoiceViewState, (state) => ({
        ...state,
        selectedInvoice: invoices[0].id,
      }));
    });
  });

  test('should select invoice when clicking', () => {
    // Arrange
    const newSelectedInvoiceIndex = 2;

    // Act
    const invoiceButton = renderResult
      .getAllByRole('button')
      .at(newSelectedInvoiceIndex + 1)!;
    fireEvent.click(invoiceButton);

    // Assert
    const newSelectedInvoice = getRecoil(invoiceViewSelectedInvoiceSelector);
    expect(newSelectedInvoice).toEqual(invoices[newSelectedInvoiceIndex]);
  });
});
