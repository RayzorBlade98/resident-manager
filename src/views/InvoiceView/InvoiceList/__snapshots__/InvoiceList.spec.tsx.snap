// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvoiceList should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-10klw3m"
  >
    <ul
      class="MuiList-root MuiList-padding css-1whj2so-MuiList-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-1m5psy8-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Neue Abrechnung
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-1m5psy8-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Februar 2023 - März 2023
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiListItemButton-root MuiListItemButton-gutters Mui-selected css-1m5psy8-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            März 2023 - April 2023
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-1m5psy8-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            April 2023 - Mai 2023
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-1m5psy8-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Mai 2023 - Juni 2023
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-1m5psy8-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Juni 2023 - Juli 2023
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </ul>
  </div>
</div>
`;
